src/core/comm/coll.cc

```diff
@@ -0,0 +1,451 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <atomic>
+#include <cstdlib>
+
+#ifndef LEGATE_USE_GASNET
+#include <stdint.h>
+#endif
+
+#include "coll.h"
+#include "legion.h"
+
+namespace legate {
+namespace comm {
+namespace coll {
+
+using namespace Legion;
+Logger log_coll("coll");
+
+#if defined(LEGATE_USE_GASNET)
+
+enum CollTag : int {
+  BCAST_TAG     = 0,
+  GATHER_TAG    = 1,
+  ALLTOALL_TAG  = 2,
+  ALLTOALLV_TAG = 3,
+  MAX_TAG       = 10,
+};
+
+static std::vector<MPI_Comm> mpi_comms;
```

Comment:
This is a list of `MPI_Comm` managed by us, not the cpu comm (`CollComm`) exposed to users. When using the `ncclCommInitRank` to create a comm, only the ranks within the comm need to call it, however, MPI is much more complicated: 1. all the ranks (even the one not in the comm) have to call the MPI communicator creation function; 2. even though MPI is thread safe, only 1 thread per node can call the MPI comm creation function. So I asked the Open MPI folks at UTK how to implement something like the `ncclCommInitRank`, they suggested me to create a list of `MPI_Comm` (each one is a copy of MPI_COMM_WORLD) during the library initialization, later, when using `collCommCreate` to create a `CollComm`, we just need to get a `MPI_Comm` from the list. Later, when the `CollComm` is finalized, we can return the `MPI_Comm` back. This makes the life much easier, and the we can use the `collCommCreate` the same as the `ncclCommInitRank`

Reply:
Yeah, but we can not do it using a single index launcher just like nccl, so for now, to make it easy, I just create several MPI_Comm during init and put them in the vector. I just do not want put too much time on it because after this release, I plan to sync with Mike's student to see if he has made any progress in the realm level collective, and I also would like to take a look at the ucx to see if we need to implement the collective without MPI. MPI is not rank-per-core friendly, and has limited number of tags which limits the max number of ranks we can run.

Reply:
I don't believe you can't do it with a single index launch, but if the plan is to switch over to UCX eventually, I agree that we should spend too much time on it.